tests/unit/core/mender-artifact_test.cpp

```diff
@@ -0,0 +1,320 @@
+extern "C" {
+/* https://github.com/google/sanitizers/wiki/AddressSanitizerFlags#run-time-flags
+* Since we want to test cases where realloc will return null because of a
+* too large size, we need to set allocator_may_return_null=1 */
+const char *
+__asan_default_options() {
+    return "allocator_may_return_null=1";
+}
+const char *
+__tsan_default_options() {
+    return "allocator_may_return_null=1";
+}
+}
+#endif
+
+class MenderArtifactTest : public ::testing::Test {
+protected:
+    vector<uint8_t> artifact_data;
+    void           *data;
+    fs::path        artifact_path;
+    fs::path        script_file_name;
+
+    vector<uint8_t> CreateArtifact(string custom_script = "") {
+        string script;
+        if (custom_script.empty()) {
+            script = R"(#! /bin/sh
+            DIRNAME=$(dirname $0)
+            echo foobar > ${DIRNAME}/testdata
+
+            mender-artifact write module-image --compression none \
+            --type unit-test -n unit-test-artifact --device-type unit-test \
+            -f ${DIRNAME}/testdata -o ${DIRNAME}/unit-test-artifact.mender || exit 1)";
+        } else {
+            script = custom_script;
+        }
+
+        fs::path tmp_dir = fs::temp_directory_path();
+        script_file_name = tmp_dir / "test-script.sh";
+
+        ofstream os(script_file_name.c_str());
+
+        os << script;
+        os.close();
+
+        chmod(script_file_name.c_str(), S_IRUSR | S_IWUSR | S_IXUSR);
+        system(script_file_name.c_str());
+
+        artifact_path = tmp_dir / "unit-test-artifact.mender";
+        ifstream file(artifact_path);
+        return vector<uint8_t>((istreambuf_iterator<char>(file)), istreambuf_iterator<char>());
+    }
+
+    void InitializeData(mender_update_module_t *update_module, mender_deployment_data_t *deployment_data, mender_artifact_download_data_t *mock_download_data) {
+        mender_create_deployment_data("id", "unit-test-artifact", &deployment_data);
+
+        update_module->artifact_type = "unit-test";
+        update_module->callbacks[MENDER_UPDATE_STATE_DOWNLOAD]
+            = [](MENDER_ARG_UNUSED mender_update_state_t state, MENDER_ARG_UNUSED mender_update_state_data_u data) { return MENDER_OK; };
+        mender_update_module_register(update_module);
+
+        mock_download_data->deployment    = deployment_data;
+        mock_download_data->update_module = update_module;
+    }
+
+    mender_update_module_t          *update_module;
+    mender_deployment_data_t        *deployment_data;
+    mender_artifact_download_data_t *mock_download_data;
+
+    void SetUp() {
+        update_module      = (mender_update_module_t *)malloc(sizeof(mender_update_module_t));
+        deployment_data    = (mender_deployment_data_t *)malloc(sizeof(mender_deployment_data_t));
+        mock_download_data = (mender_artifact_download_data_t *)malloc(sizeof(mender_artifact_download_data_t));
```

Comment:
The type-casts are not needed, the assignment takes care of them. But they don't hurt, of course. :)